src/CloudFoundry.WinDEA/StagingInstance.cs

```diff
@@ -265,34 +267,226 @@ public void GetBuildpack(StagingStartMessageRequest message, string gitPath, str
+                    {
+                        found = true;
+                    }
+                }
+                if (!found)
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        public void DownloadAdminBuildpack(StagingStartRequestAdminBuildpack adminBuildpack, string adminBuildpackDir)
+        {
+            string archiveFile = Path.Combine(adminBuildpackDir, adminBuildpack.Key + ".zip");
+            string destDir = Path.Combine(adminBuildpackDir, adminBuildpack.Key);
+
+            if (Directory.Exists(destDir))
+            {
+                Logger.Info("Skipping download for admin buildpack {0} because it was previously downloaded in {1} ", adminBuildpack.Key, destDir);
+                return;
+            }
+
+            Directory.CreateDirectory(destDir);
+
+            WebClient client = new WebClient();
+
+            Logger.Info("Downloading admin buildpack {0} in {1} from {2}", adminBuildpack.Key, destDir, adminBuildpack.Url);
+            try
+            {
+                var downloadUri = new Uri(adminBuildpack.Url);
+                client.Headers[HttpRequestHeader.Authorization] = "Basic " + System.Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes(downloadUri.UserInfo));
+                client.DownloadFile(downloadUri, archiveFile);
+
+                ZipFile.ExtractToDirectory(archiveFile, destDir);
+            }
+            catch
+            {
+                Directory.Delete(destDir, true);
+                File.Delete(archiveFile);
+
+                throw;
+            }
+            finally
+            {
+                client.Dispose();
+            }
+
+            File.Delete(archiveFile);
+
+        }
+
+        public void TryCleanupUnrackedAdminBuildpacks(StagingStartRequestAdminBuildpack[] adminBuildpacks, string adminBuildpackDir)
+        {
+            var cachedAdminBuilpacks = new DirectoryInfo(adminBuildpackDir).EnumerateDirectories().ToDictionary(di => di.Name, di => di.FullName);
+
+            var adminBuildpacksLookup = adminBuildpacks.ToLookup(ab => ab.Key);
+
+            foreach (var cachedBuildpack in cachedAdminBuilpacks)
+            {
+                if (adminBuildpacksLookup.Contains(cachedBuildpack.Key))
+                {
+                    continue;
+                }
+
+                bool signaled = false;
+                var buildpackMutex = new Mutex(false, "admin-buildpack-" + cachedBuildpack.Key);
+
+                try
+                {
+
+                    signaled = buildpackMutex.WaitOne(0);
+
+                    if (signaled)
+                    {
+                        Logger.Info("Deleting old admin buildpack {0} from {1}", cachedBuildpack.Key, cachedBuildpack.Value);
+
+                        Directory.Delete(cachedBuildpack.Value, true);
+                    }
+                }
+                finally
+                {
+                    if (signaled)
```

Comment:
The cleanup is invoked synchronously at every stating request. It will just try to get the mutex without any waiting and skip the cleanup if the buildpack is in use or downloading. Even if it will skip the cleanup for an iteration, it will get cleaned-up at the next staging request when it will be unlocked.

I tried to avoid adding extra complexity, e.g. not adding extra worker-threads that do the cleanup.